### PR TITLE
feat(api): clearing status

### LINKED ISSUE #2283: Creaate REST API for the `clearing status` and  `cleared/ToBeCleared/Total` response.
### Description

Creaate REST API for the clearing status and  cleared/ToBeCleared/Total response in the License browser page.

#### Screenshots

image

The response required in the API can be understood from the Above screenshot of the old UI

PS : I would like to work on this issue.

@GMishx @shaheemazmalmmd

### PR DESCRIPTION
**Pull request by soham4abc**
## Description

Clearing status parameter for the `/uploads/ID/licenses` endpoint.

## How to test

Send a licenses request for an upload to see the response.

## Screenshot
image

Fixes #2283

PTAL @GMishx 

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2288"></a>

### CODE REVIEW

File: src/www/ui/api/Helper/UploadHelper.php [hunk 1]
```diff
@@ -627,6 +628,31 @@ public function getUploadCopyrightList($uploadId)
...
+     * @param DbManager $dbManager DbManager object
+     * @param integer $uploadTreeTableId upload tree id of the file
+     * @param integer $groupId groupId of the user
+     * @return String String containing the Clearing status message
+*/
+
+  public function fetchClearingStatus($dbManager, $uploadTreeTableId, $groupId)
+  {
+    $clearingStatus = [];
+    $clearingStatusMsg = array("WIP",null,null,"TO_BE_DISCUSSED","IRRELEVANT","IDENTIFIED","DO_NOT_USE","NON_FUNCTIONAL");
```
Comment (GMishx) [inline src/www/ui/api/Helper/UploadHelper.php:642]: Please create object of class [`Fossology\Lib\Data\DecisionTypes`](https://github.com/fossology/fossology/blob/43924f05229e03bfa1c86eb285127f9f0b93d409/src/lib/php/Data/DecisionTypes.php#L10) instead.

File: src/www/ui/api/Helper/UploadHelper.php [hunk 2]
```diff
@@ -627,6 +628,31 @@ public function getUploadCopyrightList($uploadId)
...
+    $sql = "SELECT * FROM clearing_decision WHERE uploadtree_fk=$1 AND group_fk=$2 ORDER BY date_added DESC;";
+    $stmt = __METHOD__.'.collectMainLicenses';
+    $rows = $dbManager->getRows($sql, array($uploadTreeTableId,"$groupId"), $stmt);
+    foreach ($rows as $rowItem) {
+      $clearingStatus[]=$rowItem['decision_type'];
+    }
+    if ($clearingStatus == []) {
+      return("not_found");
+    } else {
+      return ($clearingStatusMsg[(int)$clearingStatus[0]]);
```
Comment (GMishx) [inline src/www/ui/api/Helper/UploadHelper.php:652]: And use the function [`getTypeName()`](https://github.com/fossology/fossology/blob/43924f05229e03bfa1c86eb285127f9f0b93d409/src/lib/php/Data/Types.php#L30) on object of `DecisionTypes` class.

File: src/www/ui/api/Helper/UploadHelper.php [hunk 3]
```diff
@@ -643,10 +669,10 @@ public function getUploadLicenseList($uploadId, $agents, $printContainers, $bool
     $restHelper = $container->get('helper.restHelper');
     $uploadDao = $restHelper->getUploadDao();
     $agentDao = $container->get('dao.agent');
-
+    $dbManager = $restHelper->getDbHelper()->getDbManager();
     $uploadTreeTableName = $uploadDao->getUploadtreeTableName($uploadId);
     $parent = $uploadDao->getParentItemBounds($uploadId, $uploadTreeTableName);
-
+    $groupId = Auth::getGroupId();
```
Comment (GMishx) [inline src/www/ui/api/Helper/UploadHelper.php:675]:
```suggestion
    $groupId = $restHelper->getGroupId();
```

File: src/www/ui/api/Helper/UploadHelper.php [hunk 4]
```diff
@@ -627,6 +628,31 @@ public function getUploadCopyrightList($uploadId)
...
+     * @param integer $groupId groupId of the user
+     * @return String String containing the Clearing status message
+*/
+
+  public function fetchClearingStatus($dbManager, $uploadTreeTableId, $groupId)
+  {
+    $clearingStatus = [];
+    $clearingStatusMsg = array("WIP",null,null,"TO_BE_DISCUSSED","IRRELEVANT","IDENTIFIED","DO_NOT_USE","NON_FUNCTIONAL");
+    $sql = "SELECT * FROM clearing_decision WHERE uploadtree_fk=$1 AND group_fk=$2 ORDER BY date_added DESC;";
+    $stmt = __METHOD__.'.collectMainLicenses';
```
Comment (GMishx) [inline src/www/ui/api/Helper/UploadHelper.php:644]: Does [`Fossology\Lib\Dao\ClearingDao::getFileClearings()`](https://github.com/fossology/fossology/blob/43924f05229e03bfa1c86eb285127f9f0b93d409/src/lib/php/Dao/ClearingDao.php#L159) not providing correct results?

**Review by GMishx — CHANGES_REQUESTED** [review]

File: src/www/ui/api/Helper/UploadHelper.php (hunk 3 above)
Comment (soham4abc) [reply src/www/ui/api/Helper/UploadHelper.php:675]: This is causing 500 internal server error

File: src/www/ui/api/Helper/UploadHelper.php (hunk 4 above)
Comment (soham4abc) [reply src/www/ui/api/Helper/UploadHelper.php:644]: How do I get the `itemTreeBounds` object and pass it here?

Comment (soham4abc) [reply src/www/ui/api/Helper/UploadHelper.php:644]: How do I get the `itemTreeBounds` object and pass it here?

File: src/www/ui/api/Helper/UploadHelper.php (hunk 1 above)
Comment (soham4abc) [reply src/www/ui/api/Helper/UploadHelper.php:642]: Resolved

File: src/www/ui/api/Helper/UploadHelper.php (hunk 2 above)
Comment (soham4abc) [reply src/www/ui/api/Helper/UploadHelper.php:652]: Resolved

**Comment by soham4abc** [comment]
PTAL @GMishx

File: src/www/ui/api/Helper/UploadHelper.php (hunk 4 above)
Comment (GMishx) [reply src/www/ui/api/Helper/UploadHelper.php:645]: Using [Fossology\Lib\Dao\UploadDao::getItemTreeBounds()](https://github.com/fossology/fossology/blob/43924f05229e03bfa1c86eb285127f9f0b93d409/src/lib/php/Dao/UploadDao.php#L132)

And you can cache `uploadtreetablename` using [Fossology\Lib\Dao\UploadDao::getUploadtreeTableName()](https://github.com/fossology/fossology/blob/43924f05229e03bfa1c86eb285127f9f0b93d409/src/lib/php/Dao/UploadDao.php#L270) and pass it to `fetchClearingStatus()` (which is called inside a loop).

Comment (GMishx) [reply src/www/ui/api/Helper/UploadHelper.php:645]: PTAL @soham4abc

Comment (soham4abc) [reply src/www/ui/api/Helper/UploadHelper.php:645]: Yes sir, I got this, discussed this with @shaheemazmalmmd. If you can provide me some more time we can implement both clearing status and the number of cleared files compared to total files response in the same PR itself. I have started working on it. If you allow me I will update this PR accordingly. Thank you for your review!! @GMishx

File: src/lib/php/Dao/LicenseDao.php [hunk 5]
```diff
@@ -351,7 +351,8 @@ public function getLicensesPerFileNameForAgentId(ItemTreeBounds $itemTreeBounds,
     $rgtStack = array($row['rgt']);
     $lastLft = $row['lft'];
     $path = implode('/', $pathStack);
-    $this->addToLicensesPerFileName($licensesPerFileName, $path, $row, $ignore, $clearingDecisionsForLicList);
+    $uploadTreeId = $row['uploadtree_pk'];
```
Comment (shaheemazmalmmd) [inline src/lib/php/Dao/LicenseDao.php:354]: As discussed please remove uploadtreeid from here and use itemtreebonds

Comment (soham4abc) [reply src/lib/php/Dao/LicenseDao.php:354]: I am on it. Did some testing of the function last night. Moving on to implementation

Comment (soham4abc) [reply src/lib/php/Dao/LicenseDao.php:354]: resolved

File: src/www/ui/api/Helper/UploadHelper.php (hunk 4 above)
Comment (soham4abc) [reply src/www/ui/api/Helper/UploadHelper.php]: Resolved

**Comment by soham4abc** [comment]
@GMishx @shaheemazmalmmd  this PR is updated

**Comment by soham4abc** [comment]
can we mark this PR as Work in progress?? @GMishx  @shaheemazmalmmd

**Comment by soham4abc** [comment]
PTAL @GMishx @shaheemazmalmmd

**Review by GMishx — APPROVED** [review]
Changes looks good.
Tested, working as expected.

**Comment by soham4abc** [comment]
@GMishx Thanks a lot!

**Pull request merged**